File: stack_serialize.cpp
#include "stack_serialize.h"
#include "serialize.h"
#include <iostream>

using namespace std;

void StackSerializer::saveToFile(const Stack& stack, const string& filename) {
    ofstream file(filename);
    if (!file) return;

    int count = 0;
    for (Stack::SNode* curr = stack.top; curr; curr = curr->next) {
        count++;
    }
    file << count << "\n";

    Stack::SNode* curr = stack.top;
    string* tempArr = new string[count];
    for (int i = count - 1; i >= 0; --i) {
        tempArr[i] = curr->value;
        curr = curr->next;
    }

    for (int i = 0; i < count; ++i) {
        writeStringText(file, tempArr[i]);
    }
    delete[] tempArr;
}

void StackSerializer::loadFromFile(Stack& stack, const string& filename) {
    ifstream file(filename);
    if (!file) return;

    while (!stack.isEmpty()) stack.pop();

    int count;
    file >> count;
    string dummy; getline(file, dummy);

    for (int i = 0; i < count; ++i) {
        stack.push(readStringText(file));
    }
}

void StackSerializer::saveToBinaryFile(const Stack& stack, const string& filename) {
    ofstream file(filename, ios::binary | ios::trunc);
    if (!file) return;

    int count = 0;
    for (Stack::SNode* curr = stack.top; curr; curr = curr->next) {
        count++;
    }
    file.write(reinterpret_cast<const char*>(&count), sizeof(count));

    Stack::SNode* curr = stack.top;
    string* tempArr = new string[count];
    for (int i = count - 1; i >= 0; --i) {
        tempArr[i] = curr->value;
        curr = curr->next;
    }

    for (int i = 0; i < count; ++i) {
        writeString(file, tempArr[i]);
    }
    delete[] tempArr;
}

void StackSerializer::loadFromBinaryFile(Stack& stack, const string& filename) {
    ifstream file(filename, ios::binary);
    if (!file) return;

    while (!stack.isEmpty()) stack.pop();

    int count;
    file.read(reinterpret_cast<char*>(&count), sizeof(count));
    if (file.fail()) return;

    for (int i = 0; i < count; ++i) {
        string val = readString(file);
        if (file.fail()) break;
        stack.push(val);
    }
}

File: ds/stack_serialize.go
package ds

import (
	"bufio"
	"fmt"
	"os"
)

type StackSerializer struct{}

func NewStackSerializer() *StackSerializer {
	return &StackSerializer{}
}

func (s *Stack) toSlice() []string {
	out := []string{}
	for cur := s.Top; cur != nil; cur = cur.Next {
		out = append(out, cur.Value)
	}
	return out
}

func (s *StackSerializer) SaveToFile(stack *Stack, filename string) error {
	arr := stack.toSlice()
	file, err := os.Create(filename)
	if err != nil {
		return err
	}
	defer file.Close()

	fmt.Fprintln(file, len(arr))
	for i := len(arr) - 1; i >= 0; i-- {
		WriteStringText(file, arr[i])
	}
	return nil
}

func (s *StackSerializer) LoadFromFile(stack *Stack, filename string) error {
	file, err := os.Open(filename)
	if err != nil {
		return err
	}
	defer file.Close()

	stack.Top = nil
	scanner := bufio.NewScanner(file)

	var count int
	if scanner.Scan() {
		fmt.Sscanf(scanner.Text(), "%d", &count)
	}

	for i := 0; i < count; i++ {
		val := ReadStringText(scanner)
		stack.Push(val)
	}
	return nil
}

func (s *StackSerializer) SaveToBinaryFile(stack *Stack, filename string) error {
	arr := stack.toSlice()
	file, err := os.Create(filename)
	if err != nil {
		return err
	}
	defer file.Close()

	if err := WriteInt32(file, int32(len(arr))); err != nil {
		return err
	}

	for i := len(arr) - 1; i >= 0; i-- {
		if err := WriteString(file, arr[i]); err != nil {
			return err
		}
	}
	return nil
}

func (s *StackSerializer) LoadFromBinaryFile(stack *Stack, filename string) error {
	file, err := os.Open(filename)
	if err != nil {
		return err
	}
	defer file.Close()

	stack.Top = nil

	count, err := ReadInt32(file)
	if err != nil {
		return err
	}

	for i := 0; i < int(count); i++ {
		val, err := ReadString(file)
		if err != nil {
			return err
		}
		stack.Push(val)
	}
	return nil
}